drivers/video/video_stm32_venc.c

```diff
@@ -0,0 +1,899 @@
+		 * Input buffer is dropped to not introduce latency.
+		 */
+		goto out;
+	}
+
+	/* one key frame every seconds */
+	if ((data->frame_nb % VENC_DEFAULT_FRAMERATE) == 0 || data->resync) {
+		/* if frame is the first or resync needed: set as intra coded */
+		encIn.codingType = H264ENC_INTRA_FRAME;
+	} else {
+		/* if there was a frame previously, set as predicted */
+		encIn.timeIncrement = 1;
+		encIn.codingType = H264ENC_PREDICTED_FRAME;
+	}
+
+	encIn.ipf = H264ENC_REFERENCE_AND_REFRESH;
+	encIn.ltrf = H264ENC_REFERENCE;
+
+	/* set input buffers to structures */
+	encIn.busLuma = (ptr_t)input->buffer;
+	encIn.busChromaU = (ptr_t)encIn.busLuma + data->in_fmt.width * data->in_fmt.height;
+
+	encIn.pOutBuf = (uint32_t *)output->buffer;
+	encIn.busOutBuf = (uint32_t)encIn.pOutBuf;
+	encIn.outBufSize = ROUND_UP(output->size, ALIGNMENT_INCR);
+	encOut.streamSize = 0;
+
+	ret = H264EncStrmEncode(data->encoder, &encIn, &encOut, NULL, NULL, NULL);
+	output->bytesused = encOut.streamSize;
+	LOG_DBG("output=%p, encOut.streamSize=%d", output, encOut.streamSize);
+
+	switch (ret) {
+	case H264ENC_FRAME_READY:
+		/* save stream */
+		if (encOut.streamSize == 0) {
+			/* Nothing encoded */
+			data->resync = true;
+			goto out;
+		}
+		output->bytesused = encOut.streamSize;
+		break;
+	case H264ENC_FUSE_ERROR:
+		LOG_ERR("H264EncStrmEncode error=%d", ret);
+
+		LOG_ERR("DCMIPP and VENC desync at frame %d, restart the video", data->frame_nb);
+		encoder_end(data);
+
+		ret = encoder_start(data, output);
+		if (ret) {
+			goto out;
+		}
+		break;
+	default:
+		LOG_ERR("H264EncStrmEncode error=%d", ret);
+		LOG_ERR("error encoding frame %d", data->frame_nb);
+
+		encoder_end(data);
+
+		ret = encoder_start(data, output);
+		if (ret) {
+			goto out;
+		}
+
+		data->resync = true;
+
+		ret = -EIO;
+		goto out;
+	}
+
+	data->frame_nb++;
+
+out:
+	k_fifo_put(&data->in_fifo_out, input);
+	k_fifo_put(&data->out_fifo_out, output);
+
+	return ret;
+}
+
+static int stm32_venc_set_stream(const struct device *dev, bool enable, enum video_buf_type type)
+{
+	struct stm32_venc_data *data = dev->data;
+
+	ARG_UNUSED(type);
+
+	if (!enable) {
+		/* Stop VENC */
+		encoder_end(data);
+	}
+
+	return 0;
+}
+
+static int stm32_venc_enqueue(const struct device *dev, struct video_buffer *vbuf)
+{
+	struct stm32_venc_data *data = dev->data;
+
+	if (vbuf->type == VIDEO_BUF_TYPE_INPUT) {
+		k_fifo_put(&data->in_fifo_in, vbuf);
+	} else {
+		k_fifo_put(&data->out_fifo_in, vbuf);
+	}
+
+	return encode_frame(data);
+}
+
+static int stm32_venc_dequeue(const struct device *dev, struct video_buffer **vbuf,
+			      k_timeout_t timeout)
+{
+	struct stm32_venc_data *data = dev->data;
+
+	if ((*vbuf)->type == VIDEO_BUF_TYPE_INPUT) {
+		*vbuf = k_fifo_get(&data->in_fifo_out, timeout);
+	} else {
+		*vbuf = k_fifo_get(&data->out_fifo_out, timeout);
+
+	}
+
+	if (*vbuf == NULL) {
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+ISR_DIRECT_DECLARE(stm32_venc_isr)
```

Comment:
Just for curiosity: is this performance-critical since it uses an un-wrapped ISR?